[Woo POS] Processing payment animation

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13696
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR replaces the static icon used on Processing states with an animated waves image.

I've used a ProgressView as that seemed like it had good semantic meaning.

I couldn't get all the values I needed out of Figma, so this is done by eye – comments welcome. Particularly the arc radii and lengths are the product of trial and error.

I deliberately hid the progress view from accessibility when showing on the "Show Card Reader Message" view, as it feels secondary, even misleading there. The payment _isn't_ processing, the system is waiting on user action like "Remove Card".

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Launch the app and enter POS mode
2. Connect a card reader
3. Add something to the cart and check out
4. Tap your card
5. Observe that the waves shown are animated during the processing step.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

I've tested this on an iPad running iOS 16

I tested dark mode, accessibility, and transitions.

No unit tests as it's UI only.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/cf5d1535-969d-44f5-a9d6-bfe2e1a5158c


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

joshheald committed:
13696 Use animated waves for processing payment

joshheald committed:
13696 Remove unused waves image

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/221">20.2</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
13696 Accessibility for custom progress view

joshheald commented:
Only the AppCenter upload failed, retrying the build

staskus commented on WooCommerce/Classes/POS/Presentation/Reusable Views/WavesProgressViewStyle.swift:
```diff
@@ -0,0 +1,96 @@
+            }
+        }
+        .offset(x: xOffset)
+        .frame(width: 130, height: 115)
```
I think it would be good to move some of the constants that are configurable to a `Constants` enum or default properties for easier configuration.

joshheald replied:
I considered it, but I didn't want to imply that it was actually set up for that customisation yet.

I took the approach of showing the numbers in line where they're 1-use only and we've not yet considered configuration, and using a var where there's actually a relationship between the two numbers.

If we use it elsewhere, we can make it configurable as we go

staskus commented on WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Reader Messages/PointOfSaleCardPresentPaymentProcessingMessageView.swift:
```diff
@@ -6,11 +6,8 @@ struct PointOfSaleCardPresentPaymentProcessingMessageView: View {
-                .frame(width: PointOfSaleCardPresentPaymentLayout.headerSize.width,
-                       height: PointOfSaleCardPresentPaymentLayout.headerSize.height)
+            ProgressView()
+                .progressViewStyle(WaveProgressViewStyle())
```
When we transition from `ReaderMessageView` to -> `PaymentProcesingMessageView` the progress view seems to "blink" since the previous view fades out and the next one fades in. It happens due to the default `transition(.opacity)` applied to all the cardReaderViews. 

We could fix this by explicitly setting a different transition for this view:

```
        .padding(.bottom)
        .multilineTextAlignment(.center)
        .transition(.asymmetric(insertion: .identity, removal: .opacity)) // no transition when inserting, fades out on removal
```

Result:

https://github.com/user-attachments/assets/94027f18-b141-4672-8c4e-1bf8d0d223b9

We don't have a mechanism implemented to dictate explicitly that this transition needs to be applied when going from one state to another. We could consider implementing such a mechanism in the future.

joshheald replied:
Nice, thanks.

It would be even better if it could just be _the same_ progress view so that it just keeps going, but that doesn't fit too well with our view structure. With some of the suggested animations morphing from one into the other, we may have to deal with that kind of thing in future.

joshheald replied:
Done in f1d532e – I did try to use `id(_:)`, but no success.

staskus commented on WooCommerce/Classes/POS/Presentation/Reusable Views/WavesProgressViewStyle.swift:
```diff
@@ -0,0 +1,96 @@
+    }
+}
+
+private struct WaveProgressView: View {
```
np;

```suggestion
private struct CardWaveProgressView: View {
```

joshheald replied:
Done in 352fbda

staskus commented on WooCommerce/Classes/POS/Presentation/Reusable Views/WavesProgressViewStyle.swift:
```diff
@@ -0,0 +1,96 @@
+                         endAngle: .degrees(27))
+                .trim(from: isActive ? 0.0 : 0 + inactiveInset,
+                      to: isActive ? 1.0 : 1 - inactiveInset)
+                .stroke(isActive ? Color(.wooCommercePurple(.shade60)) : Color(.wooCommercePurple(.shade40)),
```
I think we can use static colors which do not change in dark mode. Especially given the previous tap/swipe state uses a similar icon which remains static in dark mode.

```suggestion
                .stroke(isActive ? Color(.withColorStudio(.wooCommercePurple, shade: .shade60)) : Color(.wooCommercePurple(.shade40)),
```

joshheald replied:
👍

joshheald replied:
Done in c2c3906

staskus reviewed (approved):
✅ Looks good!

The progress view you implemented looks to closely resemble the designs 👍 . 

I tested the progress view and its transitions for successful payment, failed payment, dark mode, and with dynamic type.

I only left a few non-blocking comments & suggestions.

joshheald committed:
13696 Rename to use CardWave for specificity

joshheald committed:
13696 Don’t change wave colors in dark mode

joshheald committed:
13696 avoid flicker between two card reader states

Display Reader Message and Processing Payment both use the waves card progress view.

When we transition between the two, there’s a visual flicker.

This change prevents that

joshheald committed:
Merge branch 'trunk' into issue/13696-processing-payment-animation

# Conflicts:
#	WooCommerce/Classes/POS/Presentation/PointOfSaleAssets.swift
#	WooCommerce/WooCommerce.xcodeproj/project.pbxproj

joshheald committed:
13696 Remove unused processing asset definition
